lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -0,0 +1,372 @@
+
+    In other words, this manages the feature from the perspective of the Vault
+    being auto-unsealed.
+    """
+
+    AUTOUNSEAL_TOKEN_SECRET_LABEL = "vault-autounseal-token"
+
+    def __init__(
+        self,
+        tls_manager: VaultTLSManager,
+        model: Model,
+        requires: VaultAutounsealRequires,
+    ):
+        self._tls_manager = tls_manager
+        self._model = model
+        self._requires = requires
+
+    def vault_configuration_details(self) -> AutounsealConfigurationDetails | None:
+        """Return the configuration details for the vault."""
+        autounseal_details = self._requires.get_details()
+        if not autounseal_details:
+            return None
+        self._tls_manager.push_autounseal_ca_cert(autounseal_details.ca_certificate)
+        ca_cert_path = self._tls_manager.get_tls_file_path_in_workload(File.AUTOUNSEAL_CA)
+        return AutounsealConfigurationDetails(
+            autounseal_details.address,
+            autounseal_details.mount_path,
+            autounseal_details.key_name,
+            self._get_autounseal_vault_token(autounseal_details),
+            ca_cert_path,
+        )
+
+    def _get_autounseal_vault_token(self, autounseal_details: AutounsealDetails) -> str:
+        """Retrieve the auto-unseal Vault token, or generate a new one if required.
+
+        Retrieves the last used token from Juju secrets, and validates that it
+        is still valid. If the token is not valid, a new token is generated and
+        stored in the Juju secret. A valid token is returned.
+
+        Args:
+            autounseal_details: The autounseal configuration details.
+
+        Returns:
+            A periodic Vault token that can be used for auto-unseal.
+
+        """
+        external_vault = VaultClient(
+            url=autounseal_details.address,
+            ca_cert_path=self._tls_manager.get_tls_file_path_in_charm(File.AUTOUNSEAL_CA),
+        )
+        existing_token = self._get_juju_secret_field(self.AUTOUNSEAL_TOKEN_SECRET_LABEL, "token")
+        # If we don't already have a token, or if the existing token is invalid,
+        # authenticate with the AppRole details to generate a new token.
+        if not existing_token or not external_vault.authenticate(Token(existing_token)):
+            external_vault.authenticate(
+                AppRole(autounseal_details.role_id, autounseal_details.secret_id)
+            )
+            # NOTE: This is a little hacky. If the token expires, every unit
+            # will generate a new token, until the leader unit generates a new
+            # valid token and sets it in the Juju secret.
+            if self._model.unit.is_leader():
+                self._set_juju_secret(
+                    self.AUTOUNSEAL_TOKEN_SECRET_LABEL, {"token": external_vault.token}
+                )
+        return external_vault.token
+
+    def _get_juju_secret_field(self, label: str, field: str) -> str | None:
```

Comment:
Shouldn't this method belong in Juju facade?

Reply:
https://github.com/canonical/vault-k8s-operator/pull/534/commits/1ab346f30cc700a7d01eee6d821e2757552fc977